### DISCUSSION TITLE
Roadmap to Public Release

### DISCUSSION THREAD

**Discussion by xblax** [xblax/flashforge_ad5m_klipper_mod#19] [General]
I'd like to start a discussion on what is still missing before this repository should go public. It should not stay closed for too long. I would be nicer to allow everyone interested to participate in the discussions / development. The demand for this seems to be there, even though we didn't really announce the ongoing development in many places. 

What do you all think should be done? Feature-wise or documentation.

I've created thee milestones: https://github.com/xblax/flashforge_adm5_klipper_mod/milestones
- Public Beta
- Public Release
- Touch control

Some planned enhancements are already tracked as an issue https://github.com/xblax/flashforge_adm5_klipper_mod/issues but probably there is other stuff to be done or even trivial things that could be improved easily.

I will create additional issues and add them to milestones as this discussion proceeds.

**Comment by consp** [comment]
> Feature-wise 

Basic stuff, so no display, working without crashes (headcrash) or frequent 'bugs' (software). Considering the software which is running that is mostly out of our hands anyway so I would say no obvious bugs due to the way the OS is configured.

* My guess wifi would be something complained about if not present straight away. I would release it without and add as a first milestone. The issue is open already.

> documentation.

Basic idiot proof diy description with massive caveat-emptor. You already made lots of issues and discussions which should be fine. It's better than most commercial software. I would add a basic "how to build" so someone with a bit less experience embedded wise can do some basic stuff.

**Reply by jlodew0** [reply]
Same idea here. altough, I don't use wifi. But I can imagine that a lot of users do.
Maybe an instruction to change /etc/interface/network for a static ip config. And a big disclaimer that people don't uncomment the lines in the config for calibration of the loadcells. 😁

**Reply by xblax** [reply]
For Wifi it's more or less straightforward how it works. Questions is more how it can be configured. I will add basic Wifi support to public beta milestone.

**Comment by KaruroChori** [comment]
I think there are few important things to be validated, possibly by more than a single person:
- It is actually possible to revert back to the stock firmware (I never tried it myself).
- There is no evident memory leak over a span of 24h.
- The hardware limits which have been set as defaults are reasonable. Temperature protections should be in place. And since we have a load cell working I would also add something about collisions.

And few things to tweak:
- Filament runout in my opinion should be turned on by default. There is a second sensor which is shown but not in use. I guess that one should be removed.
- We should remove any UI element suggesting to upgrade to klipper 0.12.

**Reply by xblax** [reply]
Yes, good points. Also the "power" outputs must be removed from Mainsail/Fluidd, because we don't really know yet what they do. 
> We should remove any UI element suggesting to upgrade to klipper 0.12.

Not sure how to do that. With ntp timesync now enabled, Moonraker will subscribe to it's notification channels and show that in Mainsail/Fluidd (it was failing cert verification before). But any 'auto' update should not be possible already, and it would fail probably, if it was.

We could try block these subscriptions in iptables, don't know if that's nice though ...

**Reply by consp** [reply]
> It is actually possible to revert back to the stock firmware (I never tried it myself).

Works by running the uninstall (`bash /etc/init.d/S40klipper_mod uninstall`) and reflashing the original just to be sure.

Did it twice now.

Note: did a leveling just to be sure. No idea why it would be needed but it felt safe.

**Reply by KaruroChori** [reply]
Good, we should check again with the very last version before "public beta" once it is ready :D.

**Comment by jlodew0** [comment]
> Yes, good points. Also the "power" outputs must be removed from Mainsail/Fluidd, because we don't really know yet what they do.
> 
> > We should remove any UI element suggesting to upgrade to klipper 0.12.
> 
> Not sure how to do that. With ntp timesync now enabled, Moonraker will subscribe to it's notification channels and show that in Mainsail/Fluidd (it was failing cert verification before). But any 'auto' update should not be possible already, and it would fail probably, if it was.
> 
> We could try block these subscriptions in iptables, don't know if that's nice though ...

https://klipper.discourse.group/t/i-want-to-prevent-any-update-to-klipper-and-moonraker-plus-fluid/3078/3

**Comment by consp** [comment]
I can also confirm the "revert to stock on install error" works since I used the old script calling S90 instead of S00 and it reverted.

**Comment by xblax** [comment]
I uploaded a new alpha release yesterday, now also added some release details: 
https://github.com/xblax/flashforge_adm5_klipper_mod/releases/tag/v00.02-alpha

> Second alpha release for extended testing of the base features before public beta. These are the main highlights, since v00.01-alpha:
> * Improved config and macros for Klipper (including Auto Bed Leveling before first print)
> * Wifi works via [iwd](https://wiki.archlinux.org/title/iwd) - must be configured on device at the moment
> * Camera support for uvcvideo/MJPG capable webcams - stream is available on port 8080, provided by [ustreamer](https://github.com/pikvm/ustreamer)
> * Automatic time synchronization for the mod linux
> * Reworked chroot init: starts early, before any host services. chroot mod linux now has own syslogd, klogd, dbus, mdev

Please do some testing with this. Feedback would be useful especially on: 
- Does the Webcam work for you? Especially feedback from someone with the AD5M Pro would be great.
- Can the printer connect to your local Wifi, after configured?

Some ADM5 Pro feedback in general would be good, because I don't have it.

And some feedback about regular printing would be great. I have done more printer modding than actual printing, lately :sweat_smile:
I plan to also do a multiple-hour test print in the next days.

Note: I wont't have time for more development until after the weekend. May plan is to then improve the documentation a bit, make minor fixes and have everything public by the end of next week.

**Reply by consp** [reply]
Don;t know if it was my rebuild for the klipperscreen branch but the dbus user was missing, will check it in a moment. -> It's my build.

```
 [iwd]# station wlan0 show 
                                 Station: wlan0                                
--------------------------------------------------------------------------------
  Settable  Property              Value                                          
--------------------------------------------------------------------------------
            Scanning              no                                               
            State                 connected                                        
            Connected network     <special ssid>
            IPv4 address          192.168.2.64                                     
            ConnectedBss          44:d9:e7:00:00:00                                
            Frequency             5745                                             
            Security              WPA2-Personal                                    
            RSSI                  -58 dBm                                          
            TxBitrate             434000 Kbit/s                                    
```
only tested wpa2 network which is what 99% of users wil use.

**Reply by consp** [reply]
The tgz is also just a tarball, not gzipped. But  that should not be an issue for most people.

**Reply by xblax** [reply]
Yes I know. That's because the flashforge installer does not work if it is :see_no_evil: 
They do it too, for their official releases, so I just kept it compatible.

**Reply by consp** [reply]
Haha, hadn't noticed before. Any idea what the missing dbus user could be? Did you do anything special there or is it just buildroot doing things weirdly?

**Reply by xblax** [reply]
I think I had seen that issue before and thought that the passwd file was not properly re-generated in an unclean buildroot build. But that must be checked again.

**Reply by consp** [reply]
I will build after clean, see what happens then.

**Reply by consp** [reply]
Still missing, it definately tries to put it in but the mkuser script fails without an error. I added a script to check if it is present and otherwise add it manually in the buildroot_post script. You might want to cherry-pick that as well. d1cd5d2

**Reply by consp** [reply]
I've updated the klipperscreen version to v00.02 as well for those who wish to test it.

Since I also have a 5m non pro I cannot test the cam unfortunately.

**Reply by consp** [reply]
Ran a ~3h print (4mb file) with the default print settings of orcaslicer for the ff5m (300mm/s), only wifi, master branch with klipperscreen and fluidd and mainsail open:

Start (during print)
```
# free -h
              total        used        free      shared  buff/cache   available
Mem:         107.8M       76.6M        5.6M      596.0K       25.6M       27.1M
Swap:        128.0M       27.0M      101.0M

#  ps -o pid,rss,vsz,stat,comm,args
 1076  472 3100 S    syslogd          /sbin/syslogd -n
 1080  408 3100 S    klogd            /sbin/klogd -n
 1092  404 3360 S    mdev             /sbin/mdev -df
 1108  472 3404 S    dbus-daemon      dbus-daemon --system
 1114  712 3664 S    iwd              /usr/libexec/iwd
 1203    0    0 SW   RTW_CMD_THREAD   [RTW_CMD_THREAD]
 1210  560 3100 S    udhcpc           udhcpc -t1 -A3 -b -R -O search -O staticroutes -p /var/run/udhcpc.eth0.pid -i eth0 -x hostname:ADM5
 1215 4120  45m S    Xorg             /usr/bin/Xorg :0.0 vt01 -s 0 -noreset -allowMouseOpenFail
 1223  756 3100 S    ntpd             /usr/sbin/ntpd -n -S fake-hwclock -p pool.ntp.org
 1227  844 2676 S    dropbear         /usr/sbin/dropbear -R
 1236  17m  73m S    python3          /root/printer_software/klipper/klippy-env/bin/python3 /root/printer_software/klipper/klippy/klippy.py /root/printer_data/config/printer.cfg -l /root/printer_data/logs/klipper.log -a /tmp/uds
 1243  20m 331m S    python3          /root/printer_software/moonraker/moonraker-env/bin/python3 /root/printer_software/moonraker/moonraker/moonraker.py -d /root/printer_data
 1247  416 3100 S    httpd            httpd -p 4001 -f -h /root/printer_software/web/fluidd
 1251  584 3100 S    httpd            httpd -p 4000 -f -h /root/printer_software/web/mainsail
 1258  32m 133m S    python3          /root/printer_software/KlipperScreen/KlipperScreen-env/bin/python3 /root/printer_software/KlipperScreen/screen.py
 1262    0 2152 S    getty            /sbin/getty -L ttyS3 115200 vt100

```
Max top process usage: 25%

End:
```
# free -h
              total        used        free      shared  buff/cache   available
Mem:         107.8M       76.2M        4.9M      800.0K       26.7M       27.3M
Swap:        128.0M       28.0M      100.0M
 # ps -o pid,rss,vsz,stat,comm,args
 1076  636 3100 S    syslogd          /sbin/syslogd -n
 1080  620 3100 S    klogd            /sbin/klogd -n
 1092  404 3360 S    mdev             /sbin/mdev -df
 1108  420 3404 S    dbus-daemon      dbus-daemon --system
 1114  560 3664 S    iwd              /usr/libexec/iwd
 1203    0    0 SW   RTW_CMD_THREAD   [RTW_CMD_THREAD]
 1210  560 3100 S    udhcpc           udhcpc -t1 -A3 -b -R -O search -O staticroutes -p /var/run/udhcpc.eth0.pid -i eth0 -x hostname:ADM5
 1215 3532  45m S    Xorg             /usr/bin/Xorg :0.0 vt01 -s 0 -noreset -allowMouseOpenFail
 1223  740 3100 S    ntpd             /usr/sbin/ntpd -n -S fake-hwclock -p pool.ntp.org
 1227  896 2676 S    dropbear         /usr/sbin/dropbear -R
 1236  17m  73m S    python3          /root/printer_software/klipper/klippy-env/bin/python3 /root/printer_software/klipper/klippy/klippy.py /root/printer_data/config/printer.cfg -l /root/printer_data/logs/klipper.log -a /tmp/uds
 1243  19m 331m S    python3          /root/printer_software/moonraker/moonraker-env/bin/python3 /root/printer_software/moonraker/moonraker/moonraker.py -d /root/printer_data
 1247  416 3100 S    httpd            httpd -p 4001 -f -h /root/printer_software/web/fluidd
 1251  584 3100 S    httpd            httpd -p 4000 -f -h /root/printer_software/web/mainsail
 1258  28m 133m S    python3          /root/printer_software/KlipperScreen/KlipperScreen-env/bin/python3 /root/printer_software/KlipperScreen/screen.py
 1262    0 2152 S    getty            /sbin/getty -L ttyS3 115200 vt100
```
Max top process usage (idle):  10%

**Reply by xblax** [reply]
I also attempted a ~9h test print of this model https://www.thingiverse.com/thing:5403438 (main tower). The v00.02-alpha build without klipper screen. Sliced with Prusa Slicer 2.7.1 using a custom profile, results in a ~75M G-Code file.

Unfortunately, close to finish the print stopped and I was presented with this error in Klipper:
![Screenshot_2024-03-04_07-28-00](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/9802544/bffa5025-2a57-466b-be6d-da943f20e37a)

After some research, I think I have run into that bug:
https://github.com/SoftFever/OrcaSlicer/issues/804
https://klipper.discourse.group/t/clock-rollover-and-fan-commands-causes-intermittent-timer-too-close-issues/9665

Orca Slicer has since then optimized GCode generation. Prusa slicer also an [open bug](
https://github.com/prusa3d/PrusaSlicer/issues/11856) for M106 spam. But I didn't find any upstream fixes on Klipper side:

I will re-slice with Prusa Slicer 2.7.2, avoid the dynamic fan speed and repeat the print tonight.

**Reply by consp** [reply]
Did it slow down almost to a crawl just beforehand? If so: I also ran into that stock with a 2h print.

note: I sliced with the 2.x beta branch of ocra. Haven't set up prusa yet, will try that later.

**Reply by xblax** [reply]
The print was running overnight, so I didn't have a chance to observe that. But at least it disabled the heaters on error and didn't cause a fire, so it's still a win I guess to have tested the safety features. Second attempt of that print is running now, cooling only depends on layer time. Let's hope it's it's able to finish with that settings.

**Reply by consp** [reply]
I can also confirm setting 220c as bed temperature raises errors and stops the print. (switched bed/nozzel in prusaslicer)

**Reply by consp** [reply]
@xblax How did the print go?

**Reply by xblax** [reply]
It went well on Klipper side. Klipper completed the entire 9h print without any crashes or errors. Also no sign of memory leaks. Obviously the memory usage is not identical after 9h, but in the same ball-park +-2MB. Webcam and Wifi was running as well. 

As a test, I also had it running with 32 micro-steps for the X and Y stepper. That seems to make them a bit quieter.

However, the filament jammed on the spool and extrusion therefore stopped mid print. That is not noticed by the runout sensor (as expected). Printing multiple hours without filament gave me a hard time to unclog the Nozzle. In the end I hat do heat the nozzle up, hot pull the nozzle/heat break from the extruder unit so that I could apply more force with unclogging tool.

**Reply by consp** [reply]
Oof sounds rough on the hotend. At least the good news is klipper worked. I'll try 32 microsteps as well, seems like a reasonable choice.

**Reply by consp** [reply]
> Automatic time synchronization for the mod linux

on this note. Settings the timezone with `LC_TIME` and LC_ALL=C.utf8` might be a good idea if it was not already done (haven't checked).

**Reply by xblax** [reply]
Yes, but fortunately noting seems to be damaged. I don't want to change the microsteps before public beta. We should probably start a discussion thread about optimizing the default config / printer calibration for that - I have other some other changes in mind.

**Reply by xblax** [reply]
> on this note. Settings the timezone with `LC_TIME` and LC_ALL=C.utf8` might be a good idea if it was not already done (haven't checked).

Is that required? I thought C.utf8 should be compiled-in default locale, even if nothing is explicitly set. Timezone should be UTC. There is also no timezone data enabled in buildroot.

**Reply by consp** [reply]
In that case without TZ data it's pretty much a done deal and doesn't matter. Klipperscreen starts barking error messages if LANG and LC_* are not set to something sane. I'll check what it was exactly but for now the startup script sets it to `en_US.UTF8` for simplicity sake.

I doubt this will have any effect or issues on the normal release.

**Comment by AJolly** [comment]
If you revert to stock, does it change the printing time on the printer info screen?

**Reply by consp** [reply]
If you power it off it resets, no matter what you use. The RTC has no capacitor or battery for persistence. It "resets" by ntp query in the mod so times should be off by `(now - 1970/01/01)`

**Reply by xblax** [reply]
I think @AJolly is referring to the printer statistics of the Flashforge app. That should be totally unaffected by the mod. I.e., if you print via the mod, and then go back to the stock system, the statistics do no change. The stock app doesn't know anything about the printing that happens with the mod.

**Reply by AJolly** [reply]
Exactly it was wondering about the printer statistics, because I recently needed technical support, and they asked for a photo of that screen. Want to make sure we're not causing anyone reason to have their warranty denied.

**Reply by consp** [reply]
> Want to make sure we're not causing anyone reason to have their warranty denied.

Don't know where you are but I'm pretty sure I'll win that battle no matter what they say. (Woudn't be my first rodeo, then again consumer laws here have some bite, just like my credit card company's insurance.)

If you want to set it to whatever: edit `/mnt/orig_root/opt/config/Adventurer5M.json` value `general` -> `filamentCumulative`. It's in mm. All other "data" from that screen is also in there. You'd need to remount the partition with `-o remount,rw`

edit: You asked about printing time ... not meters, but the idea is the same, just a different variable.

This includes the "serial number":
![image](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/6802998/2d7a9a94-680a-4fdd-aff9-48c28af27ebd)

There is a reason I blanked out the only thing which can actually identify your purchase. Again, these modules are so plentyful, doubt they actually can.

Might be a good idea to back it up to the data partition anyway.

**Comment by xblax** [comment]
I updated the the Public Beta Milestone with some small tasks: https://github.com/xblax/flashforge_adm5_klipper_mod/issues?q=is%3Aissue+milestone%3A%22Public+Beta%22+is%3Aopen

That are mostly some easy things to improve usability. I think most critical things are solved, otherwise. Does anybody see something that should be done and is missing on that list?

Otherwise if nobody has any show-stoppers, I would plan get everything done, make a beta release and open the repo to the public. I think it would be nice if that can happen this sunday.

**Reply by consp** [reply]
> can happen this sunday.

Are you planning a "post it on r/flashforge" and let the storm commence or a "open repo and see what happens" approach?

I don't have any issues at this point. Only issue I had was the non-restart after save_config.

**Reply by xblax** [reply]
I'm a bit conflicted what's the best approach. It seems to already attract a lot of people without being public. Minimum, I will announce the start of public beta here, and close the issue:
https://github.com/g992/flashforge-ad5m-5mpro-research/issues/6

Maybe we should also drop a hint in the Flashforge official repository, so that those already looking on Github for Klipper mods will find it:
https://github.com/FlashforgeOfficial/AD5M_Series_Klipper/issues/1

I guess an announcement on Reddit could also be made, but more like "Beta launched and everyone interested can help testing at their own risk" rather then "hey everyone, go and install this great new thing!". 

Does it make sense to open a Discord Server for Flashforge AD5M Modding (in general) with a channel for this mod? @AJolly had also suggested creating a Discord earlier.

**Reply by consp** [reply]
> It seems to already attract a lot of people without being public.

Yes people find a way if the current support is lacking. Most people do not use orcaslicer or the flashforge slicer and want to stick to what they know, which is not well supported.

> drop a hint in the Flashforge official repository

I have the distinct impression they want to keep everything secret for themselves (see lackluster release of klipper code). But that could just be me. Maybe not wake sleeping dogs. If you post it on reddit it will be visible enough I think.

> Discord Server for Flashforge AD5M Modding 

Time to spin up the old IRC clients! (Or the modern equivalent in Discord). Not my cup of tea but it's what's done these days. As long as people also submit issues in the issue tracker (with details) since that's what I'll be reading :)

**Reply by xblax** [reply]
> I have the distinct impression they want to keep everything secret for themselves (see lackluster release of klipper code). But that could just be me. Maybe not wake sleeping dogs. If you post it on reddit it will be visible enough I think.

Yeah probably you are right, but they will notice this eventually if they didn't already. Anyway, installing modified software versions is what they must allow according to GPLv3 or LGPL (re-linking requirement), so there's nothing wrong if we do it even without provided instructions. 

I don't know if they would actually provide answers if asked more specific questions via their support instead of "how can I install klipper". I have not tried to ask for anything myself.

The lackluster / non-existing measures against third-party software give me hope they just don't care that much [as Bambulab does](https://blog.bambulab.com/rooted-the-good-the-bad-and-freedom-of-choice/). If they realize that the mod does indeed work and not harm "their" system, I hope they just let it be.

**Reply by consp** [reply]
> [as Bambulab does](https://blog.bambulab.com/rooted-the-good-the-bad-and-freedom-of-choice/).

Considering almost everything is based on GPL'd stuff and it the intention not to harm them it probably fine and not 1:1 comparable as there is no jailbreak involved in the strictest sense of the idea of jailbreak.

> ... from the post ...
> However, your warranty and support will not be reinstated after switching back to Bambu Lab Official Firmware.

IANAL but also pretty sure this is illegal or at least a **_very_** gray aproaching black area of law (in most of the EU at least). Hardware malfunctions are definitely not void. User error is however not covered under this so if someone does something stupid it is their own fault. This is why ceveat emptor applies to these kind of mods. A, we as developers provide no warrenty, software is as-is etc. like in GPL is there for a reason.

**Reply by KaruroChori** [reply]
> IANAL but also pretty sure this is illegal or at least a **_very_** gray aproaching black area ...

That is my understanding as well. This kind of clause would not be acceptable for printers sold in the EU common market.
However, the boundary between hardware malfunction and hardware malfunction because of a software error is hard to determine and usually left to the seller/manufacturer.

On a practical level it would be unfair forcing them to provide support and warranty services on devices they cannot diagnose, remotely access or for which in general it is not possible to determine why the hardware failure happened in he first place.

But yes, if it is clear that the hardware failure is due to the hardware itself, they are still liable regardless of any waiver or fine print they decided to add on their terms & conditions. At least this is my understanding.

**Reply by consp** [reply]
As a test I also used SuperSlicer and no issues. Wasn't expecting any but at least it has been tested now (so almost covering all slic3r variants)

Did add the G17 code to klipper as I was getting very irritated by all the error messages.

Has anyone tested with cura?

**Reply by consp** [reply]
Another minor update (from yesterday, forgot to post). I experimented a bit with the niceness and have not noted any issues while messing with the system while klipper was at -5 (no noticeable difference with -20). The 10 for klipperscreen makes it take a bit longer to boot but afterwards I can toggle with it without the system choking and no real noticeable lag, except for the expected subprocesses running like iwctl/iwlist. @xblax added smem, I'm not familiar with it but for what I can see memory is definitely the most critical and not processing power or priority (with the current settings). I guess someone in the beta will run a several 100mb big gcode file and we shall see how that goes :rofl:

**Reply by xblax** [reply]
Yes, memory usage must be looked at in more detail at some point after beta launch. The smem utility is useful, because it takes shared memory into account. Otherwise it can look like some processes take much more than they actually do.

With klipper-screen we have too look how much page trashing happens with the swap, that could be critical. Especially in combination with a Webcam being active as well. 

I also feel like there might a regression with Klipper + Moonraker memory usage somewhere. I feel like they needed less with the initial build. But we can go back to an old chroot to verify that.

I'm not to worried about large G-Codes files. They are streamed, so in theory their size should not matter at all.

**Reply by consp** [reply]
> I also feel like there might a regression with Klipper + Moonraker memory usage somewhere. I feel like they needed less with the initial build. But we can go back to an old chroot to verify that.

Some wild guesses:
* Could be load order, has anything changed there?
* Did you compile everything with -Os or -O2/3 or changed anything there?  My experience is O3 has a bit less stack space usage than -Os. Wouldn't matter too much but everything helps, same thing with lto if available. Also hardfloat takes a bit more memory than softfloat -in general- except when you are doing lots of fpu operations.

Also matters a bit when you check ofcourse.

**Reply by xblax** [reply]
Everything should be compiled with -O3, vfpv4, neon etc. I will check cross-check current memory usage against an older build at a later point. For now I think it's working fine as is.

**Comment by xblax** [comment]
I will probably later tag the v00.03-beta version and upload the release builds. I thing the still outstanding documentation does not has to be in the tag necessarily. 

@consp 
You can then rebase KlipperScreen feature branch and tag it separately. Does it makes sense to upload the KlipperScreen build to the same Github "Release", even if it's a different tag at the moment? Then it's more obvious to newcomers that there are two variants. 

I guess it will take 1 or 2 more days until everything is ready and I finally set this to public.

**Reply by consp** [reply]
> You can then rebase KlipperScreen feature branch and tag it separately

Will do that later today, editing some docs now. (wifi specifically, it's in the feature branch right now, hope I understood the klipper_mod dir on the usb drive correctly)

> Does it makes sense to upload the KlipperScreen build to the same Github "Release", even if it's a different tag at the moment. 

It should be really clear the normal one is preferred unless they know what they are doing. Would be a shame to get issues which only show up on the KS branch to blurr the normal issues (if any exist)

**Reply by xblax** [reply]
> Will do that later today, editing some docs now. (wifi specifically, it's in the feature branch right now, hope I understood the klipper_mod dir on the usb drive correctly)

Yes, I will describe the klipper_mod dir feature in the install doc. But the Wifi doc should probably mention that an iwd config file can manually created, maybe have an example myWifi.psk and mention that it can be installed via usb drive (link to the install doc).

**Reply by consp** [reply]
> iwd config file can manually created, maybe have an example myWifi.psk and mention that it can be installed via usb drive (link to the install doc).

That's mostly what I attempted, ~except for the link to the install doc~ also done. Not my speciality, writing documentation, though.

**Reply by xblax** [reply]
Ah yes, looks good! Didn't see you had already worked on it.

**Comment by xblax** [comment]
Tagged and uploaded the public beta release. The release notes are still pending:
https://github.com/xblax/flashforge_adm5_klipper_mod/releases/tag/v00.03-beta

Will also do the missing documentation (install / uninstall, maybe camera) tomorrow evening. Too tired for today :sleeping:

**Reply by consp** [reply]
I'm building now and then testing. Should be done in a few minutes. If it works without glitches I'll add the file. I also forgot how slot it is if you use the tgz file instead of the lose files 🤣

Secretly the klipperscreen alpha already contains the nfs client utils, forgot I already added them to the buildroot ... Don't expect any issues though. Actually works as well, both nfs and nfs4 mounts. Can be merged in buildroot feature branch, added to defconfig.

And the completion sound on the end of the install was a jumpscare since I was not expecting it 😐

Looks to work as expected.

**Reply by consp** [reply]
Also went and updated some of the issues to make it more nice when people see it. (And assigned the klipperscreen bugs to myself)

**Comment by drjuliankeates** [comment]
How could you add Bash support?
